src/handlers/userChangemakerPermissionsHandlers.ts

```diff
@@ -0,0 +1,151 @@
+import {
+	createOrUpdateUserChangemakerPermission,
+	removeUserChangemakerPermission,
+} from '../database';
+import {
+	isAuthContext,
+	isId,
+	isKeycloakUserId,
+	isPermission,
+	isTinyPgErrorWithQueryContext,
+	isWritableUserChangemakerPermission,
+} from '../types';
+import {
+	DatabaseError,
+	FailedMiddlewareError,
+	InputValidationError,
+} from '../errors';
+import type { Request, Response, NextFunction } from 'express';
+
+const deleteUserChangemakerPermission = (
+	req: Request,
+	res: Response,
+	next: NextFunction,
+): void => {
+	if (!isAuthContext(req)) {
+		next(new FailedMiddlewareError('Unexpected lack of auth context.'));
+		return;
+	}
+
+	const { userKeycloakUserId, changemakerId, permission } = req.params;
+	if (!isKeycloakUserId(userKeycloakUserId)) {
+		next(
+			new InputValidationError(
+				'Invalid userKeycloakUserId parameter.',
+				isKeycloakUserId.errors ?? [],
+			),
+		);
+		return;
+	}
+	if (!isId(changemakerId)) {
+		next(
+			new InputValidationError(
+				'Invalid changemakerId parameter.',
+				isId.errors ?? [],
+			),
+		);
+		return;
+	}
+	if (!isPermission(permission)) {
+		next(
+			new InputValidationError(
+				'Invalid permission parameter.',
+				isPermission.errors ?? [],
+			),
+		);
+		return;
+	}
+
+	(async () => {
+		await removeUserChangemakerPermission(
+			userKeycloakUserId,
+			changemakerId,
+			permission,
+		);
+		res.status(204).contentType('application/json').send();
+	})().catch((error: unknown) => {
+		if (isTinyPgErrorWithQueryContext(error)) {
+			next(new DatabaseError('Error deleting item.', error));
+			return;
+		}
+		next(error);
+	});
+};
+
+const putUserChangemakerPermission = (
+	req: Request,
+	res: Response,
+	next: NextFunction,
+): void => {
+	if (!isAuthContext(req)) {
```

Comment:
Is this a repeat of what can be done using `requireAuthentication` in the routes declared below in `usersRouter.ts`? Why not use `requireAuthentication` there?

Reply:
This is a necessary type guard that allows us to treat `req` as an AuthContext

Reply:
Almost all of our routes have a `requireAuthentication` middleware but this one does not. I think that's what's giving me pause.

I take "auth" in your comment to mean both authentication and authorization. We already have a middleware for requiring authentication, but we're not using it, perhaps because we are already required to use it here in the authorization middleware? And that perhaps due to TypeScript not knowing the details of Express middleware such that we can't declare this middleware as having Request be of type Request-with-AuthContext even when we already went through a `requireAuthentication` middleware?

Regardless, I think the type check is OK here for the reason you stated (needing to access an actually present `user` object on `req`), but maybe we should re-use `requireAuthentication` on the route because it's weird to not have our ubiquitous authentication middleware not checking for authentication first. In general, I think validating things exactly once is best, but when it comes to authentication and authorization, more checks might be OK, especially given the limitations of the typechecker in an express middleware context. If we could have the type checker recognize that this code can only be called from middleware that already checked this type that might be cool but it's clear enough and low cost enough to check before use here.

Aside: I realize you are calling this a non-business-logic type check, but type checks serve the purpose of serving the business too, if only indirectly, and all our coding is intended to serve the business. So I have a quibble with the phrase "business logic" in general. This is an aside because it doesn't change your point or mine either way.

I might could be convinced that the two-in-one authorization-middleware-implicitly-checks-authentication is also OK, I'm still noodling over it.

Reply:
> perhaps because we are already required to use it in the authorization middleware requireChangemakerPermission

You may have missed the saga, but yes, and we had to refactor for this reason: https://github.com/PhilanthropyDataCommons/service/issues/1210

> In general, I think validating things exactly once is best

This isn't there for validation, it is a type guard / it is for type safety.  There is no other way to achieve type safety (we could cast it, blindly, but that would not be type safe which is why we took on #1210)

> if we could have the type checker recognize that this code can only be called from middleware that already checked this type that might be cool

We cannot :(

> it's weird to not have our ubiquitous authentication middleware checking for authentication first

`requireAuthentication` is not ubiquitous: it is currently only used for endpoints that ONLY require authentication but no additional authorization -- that's the purpose of that middleware, to simply require authentication.  We also have `requireAdministratorRole` which is used for endpoints that require that additional permission, and now this is a third rule set.  Based on the names, would you not expect those pieces of middleware to imply a requirement of authorization?  If not, then I could see it as confusing.  If so, then I think it would be redundant.

Side note: Even if we do include that extra auth middleware it would not change the definition of `requireChangemakerPermission` because that will still need the type guards.

The intention of these three distinct tiers of auth requirements is to define what the rules for accessing the route is going to be.  Alternatively, we could move to a single auth middleware that has some kind of settings parameter to specify expectations, but I think at this point it is simpler to just have distinct middleware until we have a wider range of auth rules.

Reply:
> > perhaps because we are already required to use it in the authorization middleware requireChangemakerPermission
> 
> You may have missed the saga, but yes, and we had to refactor for this reason: #1210

Right, right. That makes sense.

> > In general, I think validating things exactly once is best
> 
> This isn't there for validation, it is a type guard / it is for type safety. There is no other way to achieve type safety (we could cast it, blindly, but that would not be type safe which is why we took on #1210)

Type guards, safety, and checking are a kind of validation. We don't guard types just because the compiler tells us to do so but because we want some assurance that a thing is the expected thing. I recognize that type guards/checking/etc are a limited form of validation and we want to do further validation on the contents but I don't think there's a hard distinction here. I also understand that there is a limitation within the context of typescript and express middleware with regard to types. It's OK with me to re-check types several times given the limitations.

> > if we could have the type checker recognize that this code can only be called from middleware that already checked this type that might be cool
> 
> We cannot :(

Gotcha. It is what it is.

> > it's weird to not have our ubiquitous authentication middleware checking for authentication first
> 
> `requireAuthentication` is not ubiquitous: it is currently only used for endpoints that ONLY require authentication but no additional authorization -- that's the purpose of that middleware, to simply require authentication for routes that need nothing more than authentication. We also have `requireAdministratorRole` which is used for endpoints that require that additional permission, and now this is a third rule set. Based on the names, would you not expect those pieces of middleware to imply a requirement of authorization? If not, then I could see it as confusing. If so, then I think it would be redundant.
> 
> This isn't directly relevant, but I want to note that even if we do include that extra auth middleware it would not change the definition of `requireChangemakerPermission` because that will still need the type guards.
> 
> The intention of these three distinct tiers of auth requirements is to define what the rules for accessing the route is going to be. Alternatively, we could move to a single auth middleware that has some kind of settings parameter to specify expectations, but I think at this point it is simpler to just have distinct middleware until we have a wider range of auth rules.

I'm OK with it, on reflection. My first review was pretty rushed. During my second review, I think it crystalized.

I think it's overstating it to say the handlers and middleware here are self-documenting. I think it becomes clearer after a long hard look, but I also think given the circular reference it may be worth a comment.